src/ipv4/fragments.ml

```diff
@@ -180,3 +180,38 @@ let process cache ts (packet : Ipv4_packet.t) payload =
           | Error Hole -> maybe_add_to_cache cache', None
         else
           maybe_add_to_cache cache', None
+
+(* TODO hdr.options is a Cstruct.t atm, but instead we need to parse all the
+   options, and distinguish based on the first bit -- only these with the bit
+   set should be copied into all fragments (see RFC 791) *)
+let fragment ~mtu hdr payload =
+  let rec frag1 acc hdr hdr_buf offset data_size payload =
+    let more = Cstruct.len payload > data_size in
+    let hdr' =
+      let off = offset / 8 lor (if more then 0x2000 else 0) in
+      { hdr with Ipv4_packet.off }
+    in
+    let this_payload, rest =
+      if more then Cstruct.split payload data_size else payload, Cstruct.empty
+    in
+    let payload_len = Cstruct.len this_payload in
+    Ipv4_wire.set_ipv4_csum hdr_buf 0;
+    (match Ipv4_packet.Marshal.into_cstruct ~payload_len hdr' hdr_buf with
+     | Error msg -> invalid_arg msg
```

Comment:
According to the mli docstring, this `Error msg` is returned when `hdr_buf` is "too small to represent `hdr'`. Not clear from that comment whether `~payload_len` factors in, or if that's just used to encode the expected payload length, but I assume the latter.
Do we expect this error to occur?
- If yes: Throwing an exception doesn't seem so nice :-(
- If no: Would be good to have an explanation why not (e.g. *"cannot happen because hdr_buf is always hdr_size long"*)

Reply:
this has been in the code before, and indeed it is the case that `hdr_buf` was allocated in this function with `hdr_size` as argument, so it will never fail.

Reply:
I added a comment